Sources/Signal.swift

```diff
@@ -237,9 +309,73 @@ public final class Signal<Value, Error: Swift.Error> {
 	}
 }
 
-private struct SignalState<Value, Error: Swift.Error> {
-	var observers: Bag<Signal<Value, Error>.Observer> = Bag()
-	var retainedSignal: Signal<Value, Error>?
+/// The status of a `Signal`.
+///
+/// - note: This three-case, multi-payload enum was verified to be a packed
+///         object reference on i386, x86-64, armv7 and armv8 under the Swift
+///         3.0.1 ABI. Should any changes be made to the enum, please first
+///         verify that the change would not cause the enum to spill over the
+///         size of an object reference on all platforms.
+private enum SignalStatus<Value, Error: Swift.Error> {
+	/// The `Signal` is alive.
+	case alive(SignalState<Value, Error, ()>)
+
+	/// The `Signal` has received a termination event, and is about to be
+	/// terminated.
+	case terminating(SignalState<Value, Error, Event<Value, Error>>)
+
+	/// The `Signal` has terminated.
+	case terminated
+}
+
+/// The state of a `Signal`.
+///
+/// As the amount of state would definitely span over a cache line,
+/// `SignalState` is set to be a reference type so that we can atomically
+/// update the reference instead.
+///
+/// - warning: In-place mutation should not be introduced to `SignalState`.
+///            Copy the states and create a new instance instead.
+private final class SignalState<Value, Error: Swift.Error, U> {
+	/// The observers of the `Signal`.
+	fileprivate let observers: Bag<Signal<Value, Error>.Observer>
+
+	/// A self-retaining reference. It is set when there are one or more active
+	/// observers.
+	fileprivate let retaining: Signal<Value, Error>?
+
+	fileprivate let associatedStorage: U?
```

Comment:
Now it looks like it'd be clearer to make this `let terminatingEvent: Event<Value, Error>?` and go make to a `SignalState?`. A nil state would mean terminated, a non-nil terminating event would mean terminating, and everything else would be active.

Having to states does still seem confusing.

Otherwise maybe we should aim for a straightforward fix here and handle optimizations in a subsequent PR.